File: backend/app/src/core/interfaces/types/__tests__/models.test.py
from typing import Any, Dict


  User,
  UserRole,
  UserStatus,
  CreateUserDTO,
  UpdateUserDTO,
  Resource,
  ResourceType,
  AccessLevel,
  CreateResourceDTO,
  UpdateResourceDTO,
  Permission,
  CreatePermissionDTO,
  UpdatePermissionDTO,
  Tag,
  CreateTagDTO,
  UpdateTagDTO,
  Comment,
  CreateCommentDTO,
  UpdateCommentDTO,
  Activity,
  ActivityType,
  Notification,
  NotificationType,
  NotificationStatus,
  CreateNotificationDTO,
  UpdateNotificationDTO,
  SearchResult,
  BulkOperationResult,
} from '../models'
describe('Model Types', () => {
  describe('User Types', () => {
    it('should validate User interface', () => {
      const user: User = {
        id: '1',
        email: 'test@example.com',
        username: 'testuser',
        role: 'user',
        status: 'active',
        createdAt: Date.now(),
        updatedAt: Date.now(),
        lastLoginAt: Date.now(),
      }
      expect(user).toHaveProperty('id')
      expect(user).toHaveProperty('email')
      expect(user).toHaveProperty('username')
      expect(user).toHaveProperty('role')
      expect(user).toHaveProperty('status')
      expect(['admin', 'user', 'guest'] as UserRole[]).toContain(user.role)
      expect(['active', 'inactive', 'suspended'] as UserStatus[]).toContain(user.status)
    })
    it('should validate CreateUserDTO', () => {
      const dto: CreateUserDTO = {
        email: 'test@example.com',
        username: 'testuser',
        password: 'password123',
        role: 'user',
      }
      expect(dto).toHaveProperty('email')
      expect(dto).toHaveProperty('username')
      expect(dto).toHaveProperty('password')
      if (dto.role) {
        expect(['admin', 'user', 'guest'] as UserRole[]).toContain(dto.role)
      }
    })
    it('should validate UpdateUserDTO', () => {
      const dto: UpdateUserDTO = {
        email: 'new@example.com',
        username: 'newuser',
        role: 'admin',
        status: 'inactive',
      }
      if (dto.email) {
        expect(typeof dto.email).toBe('string')
      }
      if (dto.username) {
        expect(typeof dto.username).toBe('string')
      }
      if (dto.role) {
        expect(['admin', 'user', 'guest'] as UserRole[]).toContain(dto.role)
      }
      if (dto.status) {
        expect(['active', 'inactive', 'suspended'] as UserStatus[]).toContain(dto.status)
      }
    })
  })
  describe('Resource Types', () => {
    it('should validate Resource interface', () => {
      const resource: Resource = {
        id: '1',
        name: 'test-resource',
        type: 'file',
        ownerId: '2',
        createdAt: Date.now(),
        updatedAt: Date.now(),
        parentId: '3',
        metadata: Dict[str, Any],
      }
      expect(resource).toHaveProperty('id')
      expect(resource).toHaveProperty('name')
      expect(resource).toHaveProperty('type')
      expect(resource).toHaveProperty('ownerId')
      expect(['file', 'folder', 'document', 'image'] as ResourceType[]).toContain(resource.type)
    })
    it('should validate CreateResourceDTO', () => {
      const dto: CreateResourceDTO = {
        name: 'new-resource',
        type: 'folder',
        parentId: '1',
        metadata: Dict[str, Any],
      }
      expect(dto).toHaveProperty('name')
      expect(dto).toHaveProperty('type')
      expect(['file', 'folder', 'document', 'image'] as ResourceType[]).toContain(dto.type)
    })
    it('should validate UpdateResourceDTO', () => {
      const dto: UpdateResourceDTO = {
        name: 'updated-resource',
        parentId: '2',
        metadata: Dict[str, Any],
      }
      if (dto.name) {
        expect(typeof dto.name).toBe('string')
      }
      if (dto.parentId) {
        expect(typeof dto.parentId).toBe('string')
      }
      if (dto.metadata) {
        expect(typeof dto.metadata).toBe('object')
      }
    })
  })
  describe('Permission Types', () => {
    it('should validate Permission interface', () => {
      const permission: Permission = {
        id: '1',
        resourceId: '2',
        userId: '3',
        level: 'read',
        createdAt: Date.now(),
        updatedAt: Date.now(),
      }
      expect(permission).toHaveProperty('resourceId')
      expect(permission).toHaveProperty('userId')
      expect(permission).toHaveProperty('level')
      expect(['read', 'write', 'admin'] as AccessLevel[]).toContain(permission.level)
    })
    it('should validate CreatePermissionDTO', () => {
      const dto: CreatePermissionDTO = {
        resourceId: '1',
        userId: '2',
        level: 'write',
      }
      expect(dto).toHaveProperty('resourceId')
      expect(dto).toHaveProperty('userId')
      expect(dto).toHaveProperty('level')
      expect(['read', 'write', 'admin'] as AccessLevel[]).toContain(dto.level)
    })
  })
  describe('Activity and Notification Types', () => {
    it('should validate Activity interface', () => {
      const activity: Activity = {
        id: '1',
        userId: '2',
        resourceId: '3',
        type: 'create',
        createdAt: Date.now(),
        updatedAt: Date.now(),
        metadata: Dict[str, Any],
      }
      expect(activity).toHaveProperty('userId')
      expect(activity).toHaveProperty('resourceId')
      expect(activity).toHaveProperty('type')
      expect(['create', 'update', 'delete', 'share', 'comment', 'tag'] as ActivityType[]).toContain(activity.type)
    })
    it('should validate Notification interface', () => {
      const notification: Notification = {
        id: '1',
        userId: '2',
        type: 'mention',
        status: 'unread',
        title: 'New mention',
        message: 'You were mentioned',
        createdAt: Date.now(),
        updatedAt: Date.now(),
        metadata: Dict[str, Any],
      }
      expect(notification).toHaveProperty('userId')
      expect(notification).toHaveProperty('type')
      expect(notification).toHaveProperty('status')
      expect(['mention', 'share', 'comment', 'system'] as NotificationType[]).toContain(notification.type)
      expect(['unread', 'read', 'archived'] as NotificationStatus[]).toContain(notification.status)
    })
  })
  describe('Search and Bulk Operation Types', () => {
    it('should validate SearchResult interface', () => {
      const searchResult: SearchResult<Resource> = {
        entity: Dict[str, Any],
        score: 0.95,
        highlights: Dict[str, Any],
      }
      expect(searchResult).toHaveProperty('entity')
      expect(searchResult).toHaveProperty('score')
      expect(typeof searchResult.score).toBe('number')
    })
    it('should validate BulkOperationResult interface', () => {
      const result: BulkOperationResult = {
        successful: ['1', '2'],
        failed: [{ id: '3', error: 'Not found' }],
        total: 3,
        successCount: 2,
        failureCount: 1,
      }
      expect(result).toHaveProperty('successful')
      expect(result).toHaveProperty('failed')
      expect(result).toHaveProperty('total')
      expect(result).toHaveProperty('successCount')
      expect(result).toHaveProperty('failureCount')
      expect(Array.isArray(result.successful)).toBe(true)
      expect(Array.isArray(result.failed)).toBe(true)
    })
  })
}) 

File: src/__tests__/types/models.test.ts
import {
  BaseEntity,
  FullEntity,
  Model,
  HasOne,
  HasMany,
  BelongsTo,
  BelongsToMany,
  ModelMetadata,
  ModelHooks,
  ValidationRules,
  QueryBuilder,
  User,
  UserProfile,
  Product,
  Order,
  OrderItem,
} from '@/shared/types/models';
import { describe, it, expect } from 'vitest';
import type {
  UserRole,
  UserStatus,
  CreateUserDTO,
  UpdateUserDTO,
  Resource,
  ResourceType,
  CreateResourceDTO,
  UpdateResourceDTO,
  Permission,
  AccessLevel,
  CreatePermissionDTO,
  UpdatePermissionDTO,
  Tag,
  CreateTagDTO,
  UpdateTagDTO,
  Comment,
  CreateCommentDTO,
  UpdateCommentDTO,
  Activity,
  ActivityType,
  Notification,
  NotificationType,
  NotificationStatus,
  CreateNotificationDTO,
  UpdateNotificationDTO,
  SearchResult,
  BulkOperationResult,
} from '../../types/models';

// Type assertion helper
const assertType = <T>(_value: T) => {
  expect(true).toBe(true); // Add an assertion to satisfy Vitest
};

describe('Model Types', () => {
  describe('Base Entity Types', () => {
    it('should correctly type BaseEntity', () => {
      const entity: BaseEntity = {
        id: '123',
        createdAt: new Date(),
        updatedAt: new Date(),
        isActive: true,
      };

      expect(entity.id).toBe('123');
      expect(entity.isActive).toBe(true);
      expect(entity.createdAt).toBeInstanceOf(Date);
      expect(entity.updatedAt).toBeInstanceOf(Date);
    });

    it('should correctly type FullEntity', () => {
      const entity: FullEntity = {
        id: '123',
        createdAt: new Date(),
        updatedAt: new Date(),
        isActive: true,
        isDeleted: false,
        createdBy: 'user1',
        updatedBy: 'user2',
        version: 1,
      };

      expect(entity.id).toBe('123');
      expect(entity.isDeleted).toBe(false);
      expect(entity.createdBy).toBe('user1');
      expect(entity.version).toBe(1);
    });
  });

  describe('Model Interface', () => {
    it('should correctly type Model', () => {
      interface TestEntity extends BaseEntity {
        name: string;
      }

      const model: Model<TestEntity> = {
        data: {
          id: '123',
          name: 'Test',
          createdAt: new Date(),
          updatedAt: new Date(),
          isActive: true,
        },
        isDirty: false,
        isNew: true,
        changes: {
          name: 'Updated Test',
        },
      };

      expect(model.data.name).toBe('Test');
      expect(model.isDirty).toBe(false);
      expect(model.isNew).toBe(true);
      expect(model.changes.name).toBe('Updated Test');
    });
  });

  describe('Relationship Types', () => {
    class MockEntity implements BaseEntity {
      id: string;
      createdAt: Date;
      updatedAt: Date;
      isActive: boolean;

      constructor(id: string) {
        this.id = id;
        this.createdAt = new Date();
        this.updatedAt = new Date();
        this.isActive = true;
      }
    }

    it('should correctly implement HasOne relationship', async () => {
      const hasOne: HasOne<MockEntity> = {
        async get() {
          return new MockEntity('123');
        },
        async set(value) {
          // Implementation not needed for type test
        },
      };

      const result = await hasOne.get();
      expect(result?.id).toBe('123');
    });

    it('should correctly implement HasMany relationship', async () => {
      const hasMany: HasMany<MockEntity> = {
        async get() {
          return [new MockEntity('1'), new MockEntity('2')];
        },
        async add(item) {
          // Implementation not needed for type test
        },
        async remove(item) {
          // Implementation not needed for type test
        },
        async set(items) {
          // Implementation not needed for type test
        },
      };

      const results = await hasMany.get();
      expect(results).toHaveLength(2);
      expect(results[0].id).toBe('1');
    });

    it('should correctly implement BelongsTo relationship', async () => {
      const belongsTo: BelongsTo<MockEntity> = {
        async get() {
          return new MockEntity('123');
        },
        async set(value) {
          // Implementation not needed for type test
        },
      };

      const result = await belongsTo.get();
      expect(result?.id).toBe('123');
    });

    it('should correctly implement BelongsToMany relationship', async () => {
      const belongsToMany: BelongsToMany<MockEntity> = {
        async get() {
          return [new MockEntity('1'), new MockEntity('2')];
        },
        async add(item) {
          // Implementation not needed for type test
        },
        async remove(item) {
          // Implementation not needed for type test
        },
        async set(items) {
          // Implementation not needed for type test
        },
      };

      const results = await belongsToMany.get();
      expect(results).toHaveLength(2);
      expect(results[1].id).toBe('2');
    });
  });

  describe('Model Configuration Types', () => {
    it('should correctly type ModelMetadata', () => {
      const metadata: ModelMetadata = {
        tableName: 'users',
        primaryKey: 'id',
        timestamps: true,
        softDeletes: true,
        relations: {
          profile: {
            type: 'hasOne',
            model: 'Profile',
            foreignKey: 'userId',
          },
          posts: {
            type: 'hasMany',
            model: 'Post',
            foreignKey: 'authorId',
          },
        },
      };

      expect(metadata.tableName).toBe('users');
      expect(metadata.relations.profile.type).toBe('hasOne');
      expect(metadata.relations.posts.foreignKey).toBe('authorId');
    });

    it('should correctly type ModelHooks', async () => {
      interface TestEntity extends BaseEntity {
        name: string;
      }

      const hooks: ModelHooks<TestEntity> = {
        async beforeCreate(data) {
          data.name = data.name?.toUpperCase();
        },
        async afterCreate(model) {
          model.data.isActive = true;
        },
      };

      const data: Partial<TestEntity> = { name: 'test' };
      await hooks.beforeCreate?.(data);
      expect(data.name).toBe('TEST');
    });

    it('should correctly type ValidationRules', () => {
      const rules: ValidationRules = {
        name: {
          type: 'string',
          required: true,
          min: 2,
          max: 50,
          pattern: /^[a-zA-Z\s]+$/,
        },
        age: {
          type: 'number',
          min: 0,
          max: 120,
          custom: (value: number) => value % 1 === 0,
        },
      };

      expect(rules.name.type).toBe('string');
      expect(rules.age.custom?.(25)).toBe(true);
    });

    it('should correctly type QueryBuilder', async () => {
      interface TestEntity extends BaseEntity {
        name: string;
        age: number;
      }

      const queryBuilder: QueryBuilder<TestEntity> = {
        where(field, value) {
          return this;
        },
        whereIn(field, values) {
          return this;
        },
        whereBetween(field, range) {
          return this;
        },
        orderBy(field, direction = 'asc') {
          return this;
        },
        limit(count) {
          return this;
        },
        offset(count) {
          return this;
        },
        include(relation) {
          return this;
        },
        async get() {
          return [];
        },
        async first() {
          return null;
        },
        async count() {
          return 0;
        },
      };

      const query = queryBuilder
        .where('age', 25)
        .whereIn('name', ['John', 'Jane'])
        .orderBy('name', 'desc')
        .limit(10);

      expect(query).toBeDefined();
    });
  });

  describe('Specific Entity Types', () => {
    it('should correctly type User and UserProfile', () => {
      const profile: UserProfile = {
        id: 'profile1',
        createdAt: new Date(),
        updatedAt: new Date(),
        isActive: true,
        userId: 'user1',
        firstName: 'John',
        lastName: 'Doe',
        avatarUrl: 'https://example.com/avatar.png',
        bio: 'Hello world',
      };
      const user: User = {
        id: 'user1',
        createdAt: new Date(),
        updatedAt: new Date(),
        isActive: true,
        username: 'johndoe',
        email: 'john@example.com',
        passwordHash: 'hashed',
        roles: ['user'],
        isVerified: true,
        profile,
      };
      expect(user.profile?.firstName).toBe('John');
      expect(user.isVerified).toBe(true);
    });

    it('should correctly type Product', () => {
      const product: Product = {
        id: 'prod1',
        createdAt: new Date(),
        updatedAt: new Date(),
        isActive: true,
        name: 'Widget',
        description: 'A useful widget',
        price: 19.99,
        currency: 'USD',
        inStock: true,
        categories: ['gadgets'],
        images: ['img1.png', 'img2.png'],
      };
      expect(product.name).toBe('Widget');
      expect(product.inStock).toBe(true);
    });

    it('should correctly type Order and OrderItem', () => {
      const item: OrderItem = {
        productId: 'prod1',
        quantity: 2,
        price: 19.99,
      };
      const order: Order = {
        id: 'order1',
        createdAt: new Date(),
        updatedAt: new Date(),
        isActive: true,
        userId: 'user1',
        items: [item],
        total: 39.98,
        status: 'pending',
        placedAt: new Date(),
      };
      expect(order.items[0].productId).toBe('prod1');
      expect(order.status).toBe('pending');
    });
  });

  describe('User Types', () => {
    it('should correctly type User', () => {
      const user: User = {
        id: '1',
        email: 'test@example.com',
        username: 'testuser',
        role: 'admin',
        status: 'active',
        createdAt: Date.now(),
        updatedAt: Date.now(),
        lastLoginAt: Date.now(),
      };
      assertType<User>(user);
    });

    it('should correctly type CreateUserDTO', () => {
      const createUser: CreateUserDTO = {
        email: 'test@example.com',
        username: 'testuser',
        password: 'password123',
        role: 'user',
      };
      assertType<CreateUserDTO>(createUser);
    });

    it('should correctly type UpdateUserDTO', () => {
      const updateUser: UpdateUserDTO = {
        email: 'new@example.com',
        status: 'suspended',
      };
      assertType<UpdateUserDTO>(updateUser);
    });
  });

  describe('Resource Types', () => {
    it('should correctly type Resource', () => {
      const resource: Resource = {
        id: '1',
        name: 'test.txt',
        type: 'file',
        ownerId: '1',
        createdAt: Date.now(),
        updatedAt: Date.now(),
        metadata: { size: 1024 },
      };
      assertType<Resource>(resource);
    });

    it('should correctly type CreateResourceDTO', () => {
      const createResource: CreateResourceDTO = {
        name: 'test.txt',
        type: 'file',
        metadata: { size: 1024 },
      };
      assertType<CreateResourceDTO>(createResource);
    });

    it('should correctly type UpdateResourceDTO', () => {
      const updateResource: UpdateResourceDTO = {
        name: 'renamed.txt',
        metadata: { size: 2048 },
      };
      assertType<UpdateResourceDTO>(updateResource);
    });
  });

  describe('Permission Types', () => {
    it('should correctly type Permission', () => {
      const permission: Permission = {
        id: '1',
        resourceId: '1',
        userId: '1',
        level: 'read',
        createdAt: Date.now(),
        updatedAt: Date.now(),
      };
      assertType<Permission>(permission);
    });

    it('should correctly type CreatePermissionDTO', () => {
      const createPermission: CreatePermissionDTO = {
        resourceId: '1',
        userId: '1',
        level: 'write',
      };
      assertType<CreatePermissionDTO>(createPermission);
    });

    it('should correctly type UpdatePermissionDTO', () => {
      const updatePermission: UpdatePermissionDTO = {
        level: 'admin',
      };
      assertType<UpdatePermissionDTO>(updatePermission);
    });
  });

  describe('Tag Types', () => {
    it('should correctly type Tag', () => {
      const tag: Tag = {
        id: '1',
        name: 'important',
        color: '#ff0000',
        description: 'Important items',
        createdAt: Date.now(),
        updatedAt: Date.now(),
      };
      assertType<Tag>(tag);
    });

    it('should correctly type CreateTagDTO', () => {
      const createTag: CreateTagDTO = {
        name: 'important',
        color: '#ff0000',
        description: 'Important items',
      };
      assertType<CreateTagDTO>(createTag);
    });

    it('should correctly type UpdateTagDTO', () => {
      const updateTag: UpdateTagDTO = {
        color: '#00ff00',
      };
      assertType<UpdateTagDTO>(updateTag);
    });
  });

  describe('Comment Types', () => {
    it('should correctly type Comment', () => {
      const comment: Comment = {
        id: '1',
        resourceId: '1',
        userId: '1',
        content: 'Test comment',
        createdAt: Date.now(),
        updatedAt: Date.now(),
        editedAt: Date.now(),
      };
      assertType<Comment>(comment);
    });

    it('should correctly type CreateCommentDTO', () => {
      const createComment: CreateCommentDTO = {
        resourceId: '1',
        content: 'Test comment',
        parentId: '1',
      };
      assertType<CreateCommentDTO>(createComment);
    });

    it('should correctly type UpdateCommentDTO', () => {
      const updateComment: UpdateCommentDTO = {
        content: 'Updated comment',
      };
      assertType<UpdateCommentDTO>(updateComment);
    });
  });

  describe('Activity Types', () => {
    it('should correctly type Activity', () => {
      const activity: Activity = {
        id: '1',
        userId: '1',
        resourceId: '1',
        type: 'create',
        metadata: { details: 'Created new file' },
        createdAt: Date.now(),
        updatedAt: Date.now(),
      };
      assertType<Activity>(activity);
    });

    it('should correctly type ActivityType', () => {
      const types: ActivityType[] = ['create', 'update', 'delete', 'share', 'comment', 'tag'];
      types.forEach(type => assertType<ActivityType>(type));
    });
  });

  describe('Notification Types', () => {
    it('should correctly type Notification', () => {
      const notification: Notification = {
        id: '1',
        userId: '1',
        type: 'mention',
        status: 'unread',
        title: 'New mention',
        message: 'You were mentioned in a comment',
        metadata: { commentId: '1' },
        createdAt: Date.now(),
        updatedAt: Date.now(),
      };
      assertType<Notification>(notification);
    });

    it('should correctly type CreateNotificationDTO', () => {
      const createNotification: CreateNotificationDTO = {
        userId: '1',
        type: 'mention',
        title: 'New mention',
        message: 'You were mentioned in a comment',
        metadata: { commentId: '1' },
      };
      assertType<CreateNotificationDTO>(createNotification);
    });

    it('should correctly type UpdateNotificationDTO', () => {
      const updateNotification: UpdateNotificationDTO = {
        status: 'read',
        readAt: Date.now(),
      };
      assertType<UpdateNotificationDTO>(updateNotification);
    });
  });

  describe('Search Types', () => {
    it('should correctly type SearchResult', () => {
      const searchResult: SearchResult<Resource> = {
        entity: {
          id: '1',
          name: 'test.txt',
          type: 'file',
          ownerId: '1',
          createdAt: Date.now(),
          updatedAt: Date.now(),
        },
        score: 0.95,
        highlights: {
          name: ['<em>test</em>.txt'],
        },
      };
      assertType<SearchResult<Resource>>(searchResult);
    });
  });

  describe('Bulk Operation Types', () => {
    it('should correctly type BulkOperationResult', () => {
      const result: BulkOperationResult = {
        successful: ['1', '2', '3'],
        failed: [
          { id: '4', error: 'Not found' },
          { id: '5', error: 'Permission denied' },
        ],
        total: 5,
        successCount: 3,
        failureCount: 2,
      };
      assertType<BulkOperationResult>(result);
    });
  });
});
